### ISSUE TITLE
Problème de moteur de recherche

### ISSUE THREAD

**Issue by BSibille** [covoiturage-libre/covoiturage-libre#154]
### Comportement actuel
Lorsque je fais une recherche "Paris-Bondy", le moteur me renvoie des trajets qui ne correspondent pas du tout. En fait, il semble que les recherches sur des trajets "courtes distances" ne fonctionnent pas bien. Pour Paris, c'est assez flagrant.

### Étapes pour reproduire le problème

[] Développement
[] Production

**Comment by ndelavergne** [comment]
Oui, ça paraît logique, parce qu'il y a un battement de 25 kilomètres autour des points de départ et d'arrivée, par défaut.

**Comment by AlexBillbao** [comment]
Je suggère d' abandonner ce système de rabattement 25km par défaut. 
Je suggère d'écrire des conditions de rabattement variables selon les zones géographiques.

J'ai analysé beaucoup de situations en faisant les centaines de recherches qui m'ont permis de dresser la carte de la répartition et densité des trajets à venir que j'ai publié le 12 fevrier.
J'ai crée sur cette carte un code graphique qui indique quand les recherches renvoient ou pas vers telle ou telle ville, dans tel ou tel sens. 

Dans les zones désertiques, il faut augmenter le rayon pour que les résultats ne soient pas vides.
Dans les grandes villes, il faut restreindre ce critère à 10/15 km, où le faire varier suivant que l'utilisateur cherche un trajet local, régional, ou interégional comme suggère Bastien.

Enfin, je suggère de travailler le rédactionnel pour les résultats de recherches : Lors de l'affichage de résultats rabattus, une phrase introduise l’apparition de tels résultats " Il n'y a pas de trajet partant / arrivant pour la ville saisie, mais il y a des trajets partant / arrivant de villes situées dans un rayon de xx kms autour de la ville recherchée."

**Comment by maxb29** [comment]
Si une recherche donne plusieurs fois la même annonce (cause villes étapes à moins de 25km des villes départ/arrivée recherchée), afficher uniquement celle-ci:
 ville la plus proche du départ recherché -> ville la plus proche de l'arrivée recherchée

**Comment by nicolasleger** [comment]
Peut être en lien avec https://github.com/covoiturage-libre/covoiturage-libre/issues/139

**Comment by AlexBillbao** [comment]
Qui peut renommer cette issue par : "faire varier le rayon de rabattage des trajets" ?
Par ailleurs, cette issue est "complémentaire", voire redondante avec  #188 .

@nicolasleger  oui il y a un lien avec la 139 (trajet qui apparaissent 2 x lors d'une recherche mais tu as proposé ds  #139 une solution qui règle  #139 en empêchant un double rabattage d'étapes proches, solution qui règle le pb  #139 indépendamment du lien avec le rayon de recherche.  Donc c'est bien 12 problématiques différentes, et il faudrait renommer cette issue pour que son nom soit limpide, et la "joindre" avec  #188 pour les relier sans perdre les commentaires contenus qui montrent les différents contextes de recherches à prendre en compte !

**Comment by Poussinou** [comment]
J'ai fermé la 188
Un commentaire intéressant : 

En effet je pense que cette fonctionnalité est très importante. 
Il faudrait mettre un curseur différent pour départ et arrivée, avec les possibilités suivantes : 0, 10 , 30 , 50, 150km

En plus de la possibilité de changer le rayon manuellement, ce qui serait vraiment top c'est d'ajuster automatiquement le rayon par défaut en fonction du nombre de km entre ville de départ et arrivée.
Par exemple les personnes qui recherchent un Marseille -> Lille (1000km) , peuvent être intéressées par un Avignon -> Lille. Contrairement à une personne qui recherche Marseille-> Lyon (300km) , qui ne sera pas intéressé par Avignon->Lyon.

Je verrais bien quelque chose comme cela:
-  trajet <30km : rayon recherche=10km
- 30<trajet<100 : rayon recherche=30km
- 100<trajet<500 : rayon recherche=50km
- 500<trajet : rayon recherche=150km

Condition pour avoir ce système , régler cette histoire de trajet dupliqué dans les résultats de recherche (due aux étapes dans le rayon de recherche, ca peut vite devenir genant avec des rayons de recherche étendus)

Et ajouter dans les résultats de la recherche, a combien de km de la ville recherchée est la ville affichée. 
Dans mon exemple (recherche Marseille->Lille),  affichage : Avignon **(100 km de Marseille)** -> Lille

**Comment by maxb29** [comment]
Pour simplifier la mise en place de ce "rayon variable automatique": au lieu de calculer et récupérer le nombre de km de l'itinéraire réel, "simplement" calculer la distance (vol d'oiseau) entre les 2 villes recherchées.
Et ajuster le rayon de recherche par défaut en fonction de cette distance (comme indiqué dans le com précédent)

**Comment by nicolasleger** [comment]
Excellente idée @maxb29 !
Pas compliqué à mettre en place une fois les échelles définies.

**Comment by maxb29** [comment]
J'ai fais quelques simulations, avec ca on devrait etre pas mal :
d<10km : rayon 1km
15< d <30km : rayon recherche=5km
30<d<100: rayon recherche=10km
100<d<300 : rayon recherche=30km
300<d<500 : rayon recherche=50km
500<d : rayon recherche=100km

Le plus simple étant peut-etre finalement de faire automatiquement: rayon= distance/10    ;-)

Quelques exemples avec les rayons calculés ci-dessus :
![image](https://user-images.githubusercontent.com/25374637/31738033-a27a7aea-b449-11e7-8ea8-d68f02e9d6bf.png)

![image](https://user-images.githubusercontent.com/25374637/31738073-c044ab86-b449-11e7-9bec-a3c93ea77469.png)

![image](https://user-images.githubusercontent.com/25374637/31738093-d83c0842-b449-11e7-9268-912a1cab66ae.png)

![image](https://user-images.githubusercontent.com/25374637/31738333-955a713e-b44a-11e7-9f8d-2427e9d7343e.png)

**Comment by maxb29** [comment]
Ou plutôt dans un premier temps : 
Rayon = distance / 6 
Ca correspond environ à la fourchette haute des catégories.
L'intérêt de ce facteur calculé est qu'on peut l'ajuster facilement. Commencer avec un rayon assez élevé (facteur 6) et dans quelques mois (années) lorsqu'il y aura beaucoup + de trajets sur la plateforme, le diminuer (facteur 10 par ex.).

**Comment by webplate** [comment]
Je suis complètement d'accord avec la proposition rayon = fonction(distance à vol d'oiseau, paramètres) !!
C'est élégant et reste flexible.
Reste à optimiser les paramètres :)
En effet, un rayon un peu large pour commencer peut être intéressant vu la faible quantité d'annonces.

**Comment by AlexBillbao** [comment]
Est-ce qu'il y a eu des embyons de codage pour cette fonction de variation automatique du rayon de recherche en fonction de l'éloignement des 2 villes recherchées ?

Dans tous les cas il faut prévoir de rajouter un bouton pour faire varier manuellement le rayon on demand de l'usager.

Mais cette fonction variable me parait une des priorités de ce qu'on a a coder
(154)

**Comment by maxb29** [comment]
Tout à fait d'accord, c'est une des priorités :-) 

Tant qu'a faire, mettre 2 boutons pour faire varier les rayons autour des villes de départ / arrivée

Avec les options suivantes:
- auto (par défaut : calcul du rayon auto, si l'utilisateur ne précise rien)
- 0km
- 10 km
- 30 km
- 50 km
- 100 km
- 200 km

**Comment by djakk** [comment]
Il y aura 2 boutons, un pour faire varier le rayon autour de la ville de départ, un autre pour la ville d’arrivée ? (Ça me semble indispensable : pour un Rennes - Paris on ne met que 10km de rayon pour Rennes mais pour Paris on peut mettre 50km)

**Comment by djakk** [comment]
Proposition : afficher une carte pour voir l'effet de variation du rayon autour des villes - une carte par ville, donc 2 cartes.
